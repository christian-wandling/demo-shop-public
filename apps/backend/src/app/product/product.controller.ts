import { Param } from '@nestjs/common';
import { ProductService } from './services/product.service';
import { ProductResponse } from './dtos/product-response';
import { Public } from 'nest-keycloak-connect';
import { CustomGet } from '../common/decorators/custom-get.decorator';
import { CustomController } from '../common/decorators/custom-controller.decorator';
import { ProductListResponse } from './dtos/product-list-response';

@CustomController({
  path: 'products',
  version: '1',
})
@Public()
export class ProductController {
  constructor(private readonly productsService: ProductService) {}

  @CustomGet({ res: ProductListResponse, isArray: true })
  getAllProducts(): Promise<ProductListResponse> {
    return this.productsService.all();
  }

  @CustomGet({ path: ':id', res: ProductResponse })
  getProductById(@Param('id') id: number): Promise<ProductResponse> {
    return this.productsService.find(id);
  }
}
