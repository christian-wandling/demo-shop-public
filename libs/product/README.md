# `@demo-shop/product`

A library for product management. It enables displaying and filtering the product catalog and add them to the shopping cart.

## Features

- **Product Catalog**

  - Response grid view of product items

- **Product Details**

  - Detailed product information

- **Shopping Integration**

  - One-click add to cart functionality

- **Product Filtering**

  - Realtime full text search filter
  - Supports text and category filtering

- **State Management**
  - Signal-based reactive store
  - Synchronization with backend

## API Reference

For a complete API reference, see the [Compodoc documentation](https://christian-wandling.github.io/demo-shop-public/compodoc/frontend).

## Running unit tests

Run `nx test product` to execute the unit tests.
