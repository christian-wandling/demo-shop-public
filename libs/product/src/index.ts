export * from './lib/product.routes';
export * from './lib/models/product-filter';
export * from './lib/product.facade';
export * from './lib/components/product-search/product-search.component';
