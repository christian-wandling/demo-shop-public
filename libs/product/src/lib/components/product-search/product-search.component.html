<a class="p-2.5 text-gray-400 hover:text-gray-500">
  <span class="sr-only">Search</span>
  <img ngSrc="icons/search.svg" alt="search" height="20" width="20" />
</a>
<div class="search-input relative" [class.extended]="!!productFilter().name">
  <input
    type="text"
    name="search"
    class="block w-full my-0.5 rounded-md border-0 py-1.5 px-3 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm/6"
    placeholder="Search products "
    [(ngModel)]="productFilter().name"
    (ngModelChange)="setProductNameFilter(productFilter().name)"
    (keydown.enter)="goToProductPage()" />
  @if (productFilter().name) {
  <div class="enter-icon absolute right-2 inset-y-0 flex items-center" (click)="goToProductPage()">
    <img ngSrc="icons/enter.svg" alt="enter" height="16" width="16" />
  </div>
  }
</div>
