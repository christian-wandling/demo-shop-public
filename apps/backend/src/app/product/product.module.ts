import { Module } from '@nestjs/common';
import { ProductRepository } from './services/product.repository';
import { ProductController } from './controllers/product.controller';
import { CommonModule } from '../common/common.module';
import { ProductService } from './services/product.service';

@Module({
  imports: [CommonModule],
  controllers: [ProductController],
  providers: [ProductRepository, ProductService],
})
export class ProductModule {}
