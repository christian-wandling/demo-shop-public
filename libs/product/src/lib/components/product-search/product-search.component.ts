import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { ProductFacade } from '../../product.facade';
import { Router } from '@angular/router';
import { FormsModule } from '@angular/forms';

/**
 * Component for searching products in the application.
 *
 * This standalone component provides a product search interface that allows users to filter products by name
 * and navigate to the products page. It uses OnPush change detection strategy for improved performance.
 *
 * @example
 * <lib-product-search/>
 */
@Component({
  selector: 'lib-product-search',
  standalone: true,
  imports: [CommonModule, NgOptimizedImage, FormsModule],
  templateUrl: './product-search.component.html',
  styleUrl: './product-search.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
  host: {
    class: 'flex align-middle lg:ml-6 cursor-pointer relative',
    'data-testid': 'product-search',
  },
})
export class ProductSearchComponent {
  readonly #productFacade = inject(ProductFacade);
  readonly productFilter = this.#productFacade.getFilter();
  readonly #router = inject(Router);

  /**
   * Updates the product name filter in the application state
   *
   * @param name - The product name to filter by, or undefined to clear the filter
   */
  setProductNameFilter(name?: string): void {
    this.#productFacade.setFilter({
      name,
    });
  }

  /**
   * Navigates the user to the products page
   *
   * This method is typically triggered when a user submits a search or clicks
   * on the search component.
   */
  goToProductPage(): void {
    this.#router.navigateByUrl('/products');
  }
}
