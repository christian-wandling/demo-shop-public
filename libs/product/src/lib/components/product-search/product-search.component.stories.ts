import { applicationConfig, Meta, moduleMetadata, StoryObj } from '@storybook/angular';
import { CommonModule, NgOptimizedImage } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { Router } from '@angular/router';
import { importProvidersFrom, signal } from '@angular/core';
import { ProductSearchComponent } from './product-search.component';
import { ProductFacade } from '../../product.facade';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

const productFilter = signal({ name: '' });

const mockProductFacade = {
  getFilter: () => productFilter,
  setFilter: (filter: { name: string }) => productFilter.set(filter),
};

const mockRouter = {
  navigateByUrl: () => {
    return;
  },
};

const meta: Meta<ProductSearchComponent> = {
  component: ProductSearchComponent,
  title: 'Product/ProductSearchComponent',
  decorators: [
    applicationConfig({
      providers: [
        importProvidersFrom(CommonModule, FormsModule, NgOptimizedImage),
        { provide: ProductFacade, useValue: mockProductFacade },
        { provide: Router, useValue: mockRouter },
      ],
    }),
  ],
  argTypes: {
    setProductNameFilter: { action: 'setProductNameFilter', control: false },
    goToProductPage: { action: 'goToProductPage', control: false },
  },
  parameters: {
    controls: {
      exclude: ['#productFacade', '#router', 'productFilter'],
    },
  },
};
export default meta;
type Story = StoryObj<ProductSearchComponent>;

export const Default: Story = {
  render: args => ({
    props: args,
    template: `
      <div style="width: 300px;">
        <lib-product-search></lib-product-search>
      </div>
    `,
  }),
  play: async ({ canvasElement, step }) => {
    const inputElement = within(canvasElement).getByRole('textbox');
    expect(inputElement).toHaveValue(productFilter().name);
  },
  parameters: {
    assets: ['icons/search.svg', 'icons/enter.svg'],
  },
};
