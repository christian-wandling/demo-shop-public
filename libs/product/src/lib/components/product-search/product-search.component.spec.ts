import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ProductSearchComponent } from './product-search.component';
import { ProductFacade } from '../../product.facade';
import { provideRouter, Router } from '@angular/router';
import { signal } from '@angular/core';
import { By } from '@angular/platform-browser';

describe('ProductSearchComponent', () => {
  let component: ProductSearchComponent;
  let fixture: ComponentFixture<ProductSearchComponent>;
  let productFacade: ProductFacade;
  let router: Router;

  const productFilter = signal({});

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ProductSearchComponent],
      providers: [
        {
          provide: ProductFacade,
          useValue: {
            setFilter: jest.fn(),
            getFilter: jest.fn().mockReturnValue(productFilter),
          },
        },
        provideRouter([]),
      ],
    }).compileComponents();

    productFacade = TestBed.inject(ProductFacade);
    router = TestBed.inject(Router);
    fixture = TestBed.createComponent(ProductSearchComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should set the product name filter', () => {
    component.setProductNameFilter('name');

    expect(productFacade.setFilter).toHaveBeenCalledWith({ name: 'name' });
  });

  it('should got to the products page', () => {
    jest.spyOn(router, 'navigateByUrl');
    component.goToProductPage();

    expect(router.navigateByUrl).toHaveBeenCalledWith('/products');
  });

  describe('search input', () => {
    it('should call setProductNameFilter on input change', () => {
      const input = fixture.debugElement.query(By.css('input'));
      fixture.detectChanges();

      input.nativeElement.value = 'test';
      input.nativeElement.dispatchEvent(new Event('input'));
      fixture.detectChanges();

      expect(productFacade.setFilter).toHaveBeenCalledWith({ name: 'test' });
    });

    it('should call goToProductPage when Enter is pressed', () => {
      jest.spyOn(router, 'navigateByUrl');
      const input = fixture.debugElement.query(By.css('input'));

      input.triggerEventHandler('keydown.enter', {});
      fixture.detectChanges();

      expect(router.navigateByUrl).toHaveBeenCalledWith('/products');
    });

    it('should display the enter icon when a value is present', () => {
      productFilter.set({ name: 'name' });
      fixture.detectChanges();

      const enterIcon = fixture.debugElement.query(By.css('.enter-icon'));
      expect(enterIcon).toBeTruthy();
    });

    it('should not display the enter icon when product name is empty', () => {
      productFilter.set({});
      fixture.detectChanges();

      const enterIcon = fixture.debugElement.query(By.css('.enter-icon'));
      expect(enterIcon).toBeFalsy();
    });

    it('should navigate to product page when enter icon is clicked', () => {
      jest.spyOn(router, 'navigateByUrl');
      productFilter.set({ name: 'name' });
      fixture.detectChanges();

      const enterIcon = fixture.debugElement.query(By.css('.absolute.right-2'));
      enterIcon.triggerEventHandler('click', {});
      fixture.detectChanges();

      expect(router.navigateByUrl).toHaveBeenCalledWith('/products');
    });
  });
});
