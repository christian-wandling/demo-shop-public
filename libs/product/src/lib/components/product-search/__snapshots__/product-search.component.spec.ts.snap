// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductSearchComponent should match the snapshot 1`] = `
<lib-product-search
  productFilter={[Function Function]}
>
  <a
    class="p-2.5 text-gray-400 hover:text-gray-500"
  >
    <span
      class="sr-only"
    >
      Search
    </span>
    <img
      alt="search"
      fetchpriority="auto"
      height="20"
      loading="lazy"
      ng-img="true"
      ngsrc="icons/search.svg"
      src="icons/search.svg"
      width="20"
    />
  </a><div
    class="search-input relative"
  >
    <input
      class="block w-full my-0.5 rounded-md border-0 py-1.5 px-3 text-gray-900 ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm/6 ng-untouched ng-pristine ng-valid"
      name="search"
      placeholder="Search products "
      type="text"
    />
  </div>
</lib-product-search>
`;
